3scale/porta
# THREESCALE-8782: Fix invoice period validation - do not allow incorrect years

**What this PR does / why we need it**:

After an invoice is generated (with the current month as the billing period), it is possible to change the billing period, which is accepted in format `YYYY-MM`, e.g. `2022-10`.

However, it was possible to set values such as "0007" or "abcd" (which is converted internally to "0000").

The result of this was that on the Invoice index page, querying for all available years (in order to fill the values in the year selector for search) was failing with an exception:
```
There are too many years options to be built. Are you sure you haven't mistyped something? You can provide the :max_years_allowed parameter.
```
which in its turn resulted in those pages showing `Internal error`.

This was caught by Bugsnag: https://app.bugsnag.com/3scale-networks-sl/system/errors/633dad6e8913e90009594351?event_id=633de84b0098fffa703d0000&i=sk&m=fq

**Which issue(s) this PR fixes** 

Fixes https://issues.redhat.com/browse/THREESCALE-8782

**Verification steps** 

- Go to an account > Invoices
- Generate a new invoice
- Go to the invoice edit page
- Set an invalid year
- Verify that the invoice is not updated, but the error is shown

**Special notes for your reviewer**:

This PR also includes a fix for the `Invoice#years` method. It was making a query for all available years, but it was not adding `DISTINCT` to the selector, so there was an entry for each invoice. [This commit](https://github.com/3scale/porta/pull/3073/commits/c41a521eb67b57b694ea1a82907497bd08d9561f) fixes it, and each value is returned once.
